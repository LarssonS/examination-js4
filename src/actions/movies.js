import database from '../firebase/firebase';

//ADD_MOVIE

export const addMovie = (movie) => ({
    type: 'ADD_MOVIE',
    movie 
});

export const startAddMovie = (movieData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            title = '',
            description = '',
            director = '',
            genre = '',
            filmWrapper = '',
            raiting = 0,
            createdAt = 0
        } = movieData;
        const movie = { title, description, director, genre, filmWrapper, raiting, createdAt, uid };

        return database.ref(`users/movies`).push(movie).then(() => {
            dispatch(addMovie({ ...movie }));
          });
    };
};

export const removeMovie = ({ id } = {}) => ({
    type: 'REMOVE_MOVIE',
    id
});

export const startRemoveMovie = ({ id } = {}) => {
    return(dispatch) => {
        return database.ref(`users/movies/${id}`).remove().then(() => {
            dispatch(removeMovie({ id }));
        });
    }
};

export const editMovie = ({id, updates}) => ({
    type: 'EDIT_MOVIE',
    id,
    updates
});

export const startEditMovie = ({id, updates}) => {
    return(dispatch) => {
        return database.ref(`users/movies/${id}`).update(updates).then(() => {
           dispatch(editMovie({id, updates})); 
        });
    };
};

export function fetchListenerMovies(ref, callback) {
    ref.on("child_added", (snapshot) => {
            const data = { ...snapshot.val(), id: snapshot.key };
            callback(data);
        });
};

export function removeListenerMovie(ref, callback) {
    ref.on("child_removed", (snapshot) => {
            callback(snapshot.key);
        });
};

export function changedListenerMovie(ref, callback) {
    ref.on("child_changed", (snapshot) => {
            const data = { updates: snapshot.val(), id: snapshot.key };
            callback(data);
        });
};
