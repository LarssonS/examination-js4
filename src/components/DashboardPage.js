import React from 'react';
import MovieList from './MovieList';
import MovieListFilters from './MovieListFilters';

const DashboardPage = () => (
    <div>
        <MovieListFilters />
        <MovieList />
    </div>
);

export default DashboardPage;
