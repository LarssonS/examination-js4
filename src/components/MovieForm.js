import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import { Container, Form, FormGroup, Label, Input, Button } from 'reactstrap';

export default class MovieForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            title: props.movie ? props.movie.title : '',
            description: props.movie ? props.movie.description : '',
            director: props.movie ? props.movie.director : '',
            genre: props.movie ? props.movie.genre : '',
            filmWrapper: props.movie ? props.movie.filmWrapper : '',
            raiting: props.movie ? props.movie.raiting : 0,
            createdAt: props.movie ? moment(props.movie.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({ title }));
    };
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onDirectorChange = (e) => {
        const director = e.target.value;
        this.setState(() => ({ director }));
    };
    onGenreChange = (e) => {
        const genre = e.target.value;
        this.setState(() => ({ genre }));
    };
    onFilmWrapperChange = (e) => {
        const filmWrapper = e.target.value;
        this.setState(() => ({ filmWrapper }));
    };
    onRaitingChange = (e) => {
        const raiting = e.target.value;
        if(raiting > 0 && raiting <= 5) {
            this.setState(() => ({ raiting }));
        }
    };
    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();

        if(
            !this.state.title ||
            !this.state.description ||
            !this.state.director ||
            !this.state.genre ||
            !this.state.filmWrapper ||
            !this.state.raiting ||
            !this.state.createdAt
         ) {
            this.setState(() => ({ error: 'Please fill in the blanks' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                title: this.state.title,
                description: this.state.description,
                director: this.state.director,
                genre: this.state.genre,
                filmWrapper: this.state.filmWrapper,
                raiting: this.state.raiting,
                createdAt: this.state.createdAt.valueOf()
            });
        }
    };

    render() {
        return (
            <div>
                <Form onSubmit={this.onSubmit}>
                    {this.state.error && <p>{this.state.error}</p>}
                    <FormGroup>
                        <Label>Movie title</Label>
                        <Input
                            type="text"
                            placeholder="Title"
                            autoFocus
                            value={this.state.title}
                            bsSize="lg"
                            onChange={this.onTitleChange}
                        />
                    </FormGroup>
                    <FormGroup>
                        <Label>Description</Label>
                        <Input
                            type="text"
                            placeholder="Description"
                            value={this.state.description}
                            bsSize="lg"
                            onChange={this.onDescriptionChange}
                        />
                    </FormGroup>
                    <FormGroup>
                        <Label>Director</Label>
                        <Input
                            type="text"
                            placeholder="Director"
                            value={this.state.director}
                            bsSize="lg"
                            onChange={this.onDirectorChange}
                        />
                    </FormGroup>
                    <FormGroup>
                        <Label>Movie genre</Label>
                        <Input type="select" bsSize="lg" onChange={this.onGenreChange}>
                            <option disabled>Select genre</option>
                            <option>None</option>
                            <option value="action">Action</option>
                            <option value="comedy">Comedy</option>
                            <option value="sci-fi">Sci-Fi</option>
                            <option value="war">War</option>
                            <option value="drama">Drama</option>
                            <option value="sport">Sport</option>
                            <option value="documentary">Documentary</option>
                        </Input>
                    </FormGroup>
                    <FormGroup>
                        <Label>Film image address</Label>
                        <Input
                            type="text"
                            placeholder="Film wrapper"
                            value={this.state.filmWrapper}
                            bsSize="lg"
                            onChange={this.onFilmWrapperChange}
                        />
                    </FormGroup>
                    <FormGroup>
                        <Label>Movie raiting</Label>
                        <Input
                            type="number" min="1" max="5"
                            placeholder="Rate"
                            value={this.state.raiting}
                            bsSize="lg"
                            onChange={this.onRaitingChange}
                        />
                    </FormGroup>
                    <FormGroup>
                        <SingleDatePicker
                            date={this.state.createdAt}
                            onDateChange={this.onDateChange}
                            focused={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                        />
                    </FormGroup>
                    <FormGroup>
                        {this.props.movie ? <Button size="lg">Update movie</Button> : <Button size="lg">Add Movie</Button>}
                    </FormGroup>
                </Form>
            </div>
        )
    }
}