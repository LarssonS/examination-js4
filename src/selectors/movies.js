import moment from 'moment';

//Get visible movies
export default (movies, { text, sortBy, startDate, endDate }) => {
    return movies.filter((movie) => {
        const createdAtMoment = moment(movie.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const titleMatch = movie.title.toLowerCase().includes(text.toLowerCase());
        const genreMatch = movie.genre.toLowerCase().includes(sortBy.toLowerCase());

        return startDateMatch && endDateMatch && titleMatch && genreMatch;
    });
}