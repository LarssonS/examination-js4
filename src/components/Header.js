import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';
import {
  Navbar,
  Nav,
  NavItem,
  Button,
  Container
} from 'reactstrap';

export const Header = ({ startLogout }) => (
    <Navbar color="light" light expand="md">
        <Container>
            <Link to="/dashboard" className="text-dark h1">
                MovieHub
            </Link>
                <Nav className="ml-auto" navbar>
                    <NavItem className="mr-5">
                        <Link to="/create" className="text-dark h2">
                            Add a movie
                        </Link>
                    </NavItem>
                    <NavItem>
                        <Button className="btn-lg" onClick={startLogout}>Logout</Button>
                    </NavItem>
                </Nav>
        </Container>
    </Navbar>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);
