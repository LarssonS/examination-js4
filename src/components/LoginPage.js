import React from 'react';
import { connect } from 'react-redux';
import { startLogin, startOnLogin, startOnRegister } from '../actions/auth';
import { Button, Form, FormGroup, Label, Input, Jumbotron, Container,
        Navbar, NavbarToggler, NavbarBrand, Nav, NavItem, NavLink, Collapse } from 'reactstrap';

export class LoginPage extends React.Component {
    constructor(props) {
        super(props);
        
        this.showRegisterForm = this.showRegisterForm.bind(this);
        this.showLoginForm = this.showLoginForm.bind(this);

        this.state = {
            register: false,
            login: false,
            isOpen: false
        };
    }
    toggle = () => {
        this.setState(() => ({
            isOpen: !this.state.isOpen
        }));
    }
    startLogin = () => {
        this.props.startLogin();
    }
    showRegisterForm = () => {
        if (this.state.register === false) {
            this.setState(() => ({
                register: true,
                login: false
            }));
        } else {
            this.setState(() => ({
                register: false
            }));
        }
    }
    showLoginForm = () => {
        if (this.state.login === false) {
            this.setState(() => ({
                register: false,
                login: true
            }));
        } else {
            this.setState(() => ({
                login: false
            }));
        }
    }
    onRegister = (e) => {
        e.preventDefault();
        const email = e.target.email.value;
        const password = e.target.password.value;

        startOnRegister(email, password);
    }
    onLogin = (e) => {
        e.preventDefault();
        const email = e.target.email.value;
        const password = e.target.password.value;

        startOnLogin(email, password);
    }
    render() {
        return (
            <div>
                <Navbar color="light" light expand="md">
                    <Container>
                        <NavbarBrand>
                            <h1>MovieHub</h1>
                        </NavbarBrand>
                        <NavbarToggler onClick={this.toggle} />
                        <Collapse isOpen={this.state.isOpen} navbar>
                            <Nav className="ml-auto" navbar>
                                <NavItem className="ml-2">
                                    <Button className="btn-lg" onClick={this.showRegisterForm}>Register</Button>
                                </NavItem>
                                <NavItem className="ml-2">
                                    <Button className="btn-lg" onClick={this.showLoginForm}>Login</Button>
                                </NavItem>
                                <NavItem className="ml-2">
                                    <Button className="btn-lg" onClick={this.startLogin}>Login with Google</Button>
                                </NavItem>
                            </Nav>
                        </Collapse>
                    </Container>    
                </Navbar>

                {this.state.register && <Container className="mt-5"> <Form onSubmit={this.onRegister}>
                    <FormGroup>
                        <Label>Email</Label>
                        <Input type="text" name="email" className="form-control-lg" placeholder="something@email.com" />
                    </FormGroup>
                    <FormGroup>
                    <Label>Password</Label>
                        <Input type="password" name="password" className="form-control-lg" placeholder="Your secret" />
                    </FormGroup>      
                    <Button>Register</Button>
                </Form> </Container>}

                {this.state.login && <Container className="mt-5"> <form onSubmit={this.onLogin}>
                    <FormGroup>
                        <Label>Email</Label>
                        <Input type="text" name="email" className="form-control-lg" placeholder="something@email.com" />
                    </FormGroup>
                    <FormGroup>
                        <Label>Password</Label>
                        <Input type="password" name="password" className="form-control-lg" placeholder="Your secret" />
                    </FormGroup>
                    <Button>Login</Button>
                </form> </Container>}

                <Jumbotron className="h-100">
                    <Container className="w-50">
                        <h1 className="display-3">Welcome to MovieHub!</h1>
                        <h2>This is a simple movie blog!</h2>
                        <hr className="my-2" />
                        <p className="h2">Register and login. Or sign in with google!</p>
                    </Container>
                </Jumbotron>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
