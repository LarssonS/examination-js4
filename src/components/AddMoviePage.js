import React from 'react';
import { connect } from 'react-redux';
import MovieForm from './MovieForm';
import { startAddMovie } from '../actions/movies';
import { Container } from 'reactstrap';

export class AddMoviePage extends React.Component {
    onSubmit = (movie) => {
        this.props.startAddMovie(movie);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <Container>
                    <h1>Add Movie</h1>
                    <MovieForm
                        onSubmit={this.onSubmit}
                    />
                </Container>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddMovie: (movie) => dispatch(startAddMovie(movie))
});

export default connect(undefined, mapDispatchToProps)(AddMoviePage);