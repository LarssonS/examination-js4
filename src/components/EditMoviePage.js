import React from 'react';
import { connect } from 'react-redux';
import MovieForm from './MovieForm';
import { startEditMovie, startRemoveMovie } from '../actions/movies';
import { Container, Button } from 'reactstrap';

export class EditMoviePage extends React.Component {
    onSubmit = (movie) => {
        this.props.startEditMovie(this.props.movie.id, movie),
        this.props.history.push('/');
    };
    onRemove = () => {
        this.props.startRemoveMovie({ id: this.props.movie.id }),
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <Container>
                    <MovieForm 
                        movie={this.props.movie}
                        onSubmit={this.onSubmit}
                    />
                    <Button onClick={this.onRemove}>Remove</Button>
                </Container>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        movie: state.movies.find((movie) => movie.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch) => ({
    startEditMovie: (id, movie) => dispatch(startEditMovie(id, movie)),
    startRemoveMovie: (data) => dispatch(startRemoveMovie(data))
});


export default connect(mapStateToProps, mapDispatchToProps)(EditMoviePage)