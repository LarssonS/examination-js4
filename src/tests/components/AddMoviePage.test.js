import React from 'react';
import { shallow } from 'enzyme';
import { AddMoviePage } from '../../components/AddMoviePage';
import movies from '../fixtures/movies';

let startAddMovie, history, wrapper;

beforeEach(() => {
  startAddMovie = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<AddMoviePage startAddMovie={startAddMovie} history={history} />);
});

test('should render AddMoviePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  wrapper.find('MovieForm').prop('onSubmit')(movies[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startAddMovie).toHaveBeenLastCalledWith(movies[1]);
});
