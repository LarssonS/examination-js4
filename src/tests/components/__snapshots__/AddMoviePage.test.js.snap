// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddMoviePage correctly 1`] = `
<div>
  <Container
    tag="div"
  >
    <h1>
      Add Movie
    </h1>
    <MovieForm
      onSubmit={[Function]}
    />
  </Container>
</div>
`;
